File: src/main/java/com/wooteco/wiki/document/service/DocumentService.java
package com.wooteco.wiki.document.service;

import com.wooteco.wiki.document.domain.CrewDocument;
import com.wooteco.wiki.document.domain.Document;
import com.wooteco.wiki.document.domain.dto.DocumentCreateRequest;
import com.wooteco.wiki.document.domain.dto.DocumentResponse;
import com.wooteco.wiki.document.domain.dto.DocumentUpdateRequest;
import com.wooteco.wiki.document.domain.dto.DocumentUuidResponse;
import com.wooteco.wiki.document.repository.DocumentRepository;
import com.wooteco.wiki.global.common.PageRequestDto;
import com.wooteco.wiki.global.exception.ErrorCode;
import com.wooteco.wiki.global.exception.WikiException;
import com.wooteco.wiki.history.service.HistoryService;
import com.wooteco.wiki.organizationdocument.dto.response.OrganizationDocumentResponse;
import com.wooteco.wiki.organizationdocument.service.DocumentOrganizationLinkService;
import java.time.LocalDateTime;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import kotlin.random.Random;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
public class DocumentService {

    private final DocumentRepository documentRepository;
    private final DocumentOrganizationLinkService organizationDocumentLinkService;
    private final HistoryService historyService;
    private final Random random;

    public DocumentService(
            DocumentRepository documentRepository,
            DocumentOrganizationLinkService organizationDocumentLinkService,
            HistoryService historyService,
            Random random
    ) {
        this.documentRepository = documentRepository;
        this.organizationDocumentLinkService = organizationDocumentLinkService;
        this.historyService = historyService;
        this.random = random;
    }

    public DocumentResponse post(DocumentCreateRequest request) {
        String title = request.getTitle();
        String contents = request.getContents();
        String writer = request.getWriter();
        Long documentBytes = request.getDocumentBytes();
        UUID uuid = request.getUuid();

        if (documentRepository.existsByTitle(title)) {
            throw new WikiException(ErrorCode.DOCUMENT_DUPLICATE);
        }

        Document document = new CrewDocument(
                title,
                contents,
                writer,
                documentBytes,
                uuid
        );

        Document savedDocument = documentRepository.save(document);
        historyService.save(savedDocument);
        return mapToResponse(savedDocument);
    }

    public DocumentResponse getRandom() {
        List<Document> documents = documentRepository.findAll();
        if (documents.isEmpty()) {
            throw new WikiException(ErrorCode.DOCUMENT_NOT_FOUND);
        }
        Document document = documents.get(random.nextInt(documents.size()));
        return mapToResponse(document);
    }

    public Page<Document> findAll(PageRequestDto requestDto) {
        return documentRepository.findAll(requestDto.toPageable());
    }

    public DocumentResponse get(String title) {
        Document document = documentRepository.findByTitle(title)
                .orElseThrow(() -> new WikiException(ErrorCode.DOCUMENT_NOT_FOUND));
        return mapToResponse(document);
    }

    public DocumentUuidResponse getUuidByTitle(String title) {
        return documentRepository.findUuidByTitle(title)
                .map(DocumentUuidResponse::new)
                .orElseThrow(() -> new WikiException(ErrorCode.DOCUMENT_NOT_FOUND));
    }

    public DocumentResponse getByUuid(UUID uuid) {
        Document document = documentRepository.findByUuid(uuid)
                .orElseThrow(() -> new WikiException(ErrorCode.DOCUMENT_NOT_FOUND));
        return mapToResponse(document);
    }

    public DocumentResponse put(UUID uuid, DocumentUpdateRequest request) {
        String title = request.getTitle();
        String contents = request.getContents();
        String writer = request.getWriter();
        Long documentBytes = request.getDocumentBytes();

        Document document = documentRepository.findByUuid(uuid)
                .orElseThrow(() -> new WikiException(ErrorCode.DOCUMENT_NOT_FOUND));

        Document updateData = document.update(title, contents, writer, documentBytes, LocalDateTime.now());
        historyService.save(updateData);
        return mapToResponse(document);
    }

    public void deleteById(Long id) {
        documentRepository.findById(id)
                .orElseThrow(() -> new WikiException(ErrorCode.DOCUMENT_NOT_FOUND));
        documentRepository.deleteById(id);
    }

    public void flushViews(Map<UUID, Integer> views) {
        List<Document> documents = documentRepository.findAllByUuidIn(views.keySet());

        for (Document document : documents) {
            Integer countToAdd = views.get(document.getUuid());
            if (countToAdd == null) {
                continue;
            }
            document.changeViewCount(document.getViewCount() + countToAdd);
        }

        documentRepository.saveAll(documents);
    }

    private DocumentResponse mapToResponse(Document document) {
        long latestVersion = historyService.findLatestVersionByDocument(document);
        List<OrganizationDocumentResponse> organizationDocumentResponses =
                (document instanceof CrewDocument crew)
                        ? organizationDocumentLinkService.findOrganizationDocumentResponsesByDocument(crew)
                        : Collections.emptyList();

        return new DocumentResponse(
                document.getId() != null ? document.getId() :
                        throwNotFound(),
                document.getUuid(),
                document.getTitle(),
                document.getContents(),
                document.getWriter(),
                document.getGenerateTime(),
                document.getViewCount(),
                latestVersion,
                organizationDocumentResponses
        );
    }

    private Long throwNotFound() {
        throw new WikiException(ErrorCode.DOCUMENT_NOT_FOUND);
    }
}

File: src/main/java/com/wooteco/wiki/document/service/DocumentService.kt
package com.wooteco.wiki.document.service

import com.wooteco.wiki.document.domain.CrewDocument
import com.wooteco.wiki.document.domain.dto.DocumentCreateRequest
import com.wooteco.wiki.document.domain.dto.DocumentResponse
import com.wooteco.wiki.document.domain.dto.DocumentUpdateRequest
import com.wooteco.wiki.document.domain.dto.DocumentUuidResponse
import com.wooteco.wiki.document.repository.DocumentRepository
import com.wooteco.wiki.global.common.PageRequestDto
import com.wooteco.wiki.global.exception.ErrorCode
import com.wooteco.wiki.global.exception.WikiException
import com.wooteco.wiki.log.service.LogService
import com.wooteco.wiki.organizationdocument.service.DocumentOrganizationLinkService
import org.springframework.data.domain.Page
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.*
import kotlin.random.Random

@Service
@Transactional
class DocumentService(
    private val documentRepository: DocumentRepository,
    private val organizationDocumentLinkService: DocumentOrganizationLinkService,
    private val logService: LogService,
    private val random: Random,
) {

    fun post(request: DocumentCreateRequest): DocumentResponse {
        val (title, contents, writer, documentBytes, uuid) = request

        if (documentRepository.existsByTitle(title)) {
            throw WikiException(ErrorCode.DOCUMENT_DUPLICATE)
        }

        val crewDocument = CrewDocument(null, title, contents, writer, documentBytes, LocalDateTime.now(), uuid)

        val savedDocument = documentRepository.save(crewDocument)
        logService.save(savedDocument)
        return mapToResponse(savedDocument);
    }

    fun getRandom(): DocumentResponse {
        val documents = documentRepository.findAll()
        if (documents.isEmpty()) {
            throw WikiException(ErrorCode.DOCUMENT_NOT_FOUND)
        }
        val document = documents[random.nextInt(documents.size)]
        return mapToResponse(document);
    }

    fun findAll(requestDto: PageRequestDto): Page<CrewDocument> {
        val pageable = requestDto.toPageable()
        return documentRepository.findAll(pageable)
    }

    fun get(title: String): DocumentResponse {
        val document = documentRepository.findByTitle(title)
            .orElseThrow { WikiException(ErrorCode.DOCUMENT_NOT_FOUND) }
        return mapToResponse(document);
    }

    fun getUuidByTitle(title: String): DocumentUuidResponse =
        documentRepository.findUuidByTitle(title)
            .map(::DocumentUuidResponse)
            .orElseThrow { WikiException(ErrorCode.DOCUMENT_NOT_FOUND) }

    fun getByUuid(uuid: UUID): DocumentResponse {
        val document = documentRepository.findByUuid(uuid)
            .orElseThrow { WikiException(ErrorCode.DOCUMENT_NOT_FOUND) }
        return mapToResponse(document);
    }

    fun put(uuid: UUID, request: DocumentUpdateRequest): DocumentResponse {
        val (title, contents, writer, documentBytes) = request

        val document = documentRepository.findByUuid(uuid)
            .orElseThrow { WikiException(ErrorCode.DOCUMENT_NOT_FOUND) }

        val updateData = document.update(title, contents, writer, documentBytes, LocalDateTime.now())

        logService.save(updateData)
        return mapToResponse(document);
    }

    fun deleteById(id: Long) {
        documentRepository.findById(id)
            .orElseThrow { WikiException(ErrorCode.DOCUMENT_NOT_FOUND) }
        documentRepository.deleteById(id)
    }

    fun flushViews(views: Map<UUID, Int>) {
        val documents = documentRepository.findAllByUuidIn(views.keys)

        for (document in documents) {
            val countToAdd = views[document.uuid] ?: continue
            document.viewCount += countToAdd
        }

        documentRepository.saveAll(documents)
    }

    private fun mapToResponse(crewDocument: CrewDocument): DocumentResponse {
        val latestVersion = logService.findLatestVersionByDocument(crewDocument);
        val organizationDocumentResponses =
            organizationDocumentLinkService.findOrganizationDocumentResponsesByDocument(crewDocument)

        return DocumentResponse(
            crewDocument.id ?: throw WikiException(ErrorCode.DOCUMENT_NOT_FOUND),
            crewDocument.uuid,
            crewDocument.title,
            crewDocument.contents,
            crewDocument.writer,
            crewDocument.generateTime,
            crewDocument.viewCount,
            latestVersion,
            organizationDocumentResponses
        )
    }
}
